### PR TITLE
feat: add task to update program skills through EMSI api

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-2904](https://2u-internal.atlassian.net/browse/PROD-2904)
Adds tasks which will be triggered through discovery on program update. Each task will then trigger the EMSI api to get the recent program skills based on program overview field. These skills will then be updated in `ProgramSkills` model.

We're adding a custom migration to recreate the `ProgramSkills` model as we wanted to change the pk constraint from a field to an autogenerated id. For some reason, we were getting some unknown errors for in case of altering the table.

**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
- [x] [Version](https://github.com/edx/taxonomy-connector/blob/master/taxonomy/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/taxonomy-connector/blob/master/CHANGELOG.rst) record added

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/taxonomy-connector/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/taxonomy-connector/actions), verify version has been pushed to [PyPI](https://pypi.org/project/taxonomy-connector/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [course-discovery](https://github.com/edx/course-discovery) to upgrade dependencies (including taxonomy-connector)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in course-discovery will look for the latest version in PyPi.

### CODE REVIEW

File: taxonomy/migrations/0021_recreate_programskill.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+                ('skill', models.ForeignKey(help_text='The ID of the skill extracted for the program.', on_delete=django.db.models.deletion.CASCADE, to='taxonomy.skill')),
+            ],
+            options={
+                'verbose_name': 'Program Skill',
+                'verbose_name_plural': 'Program Skills',
+                'ordering': ('created',),
+                'unique_together': {('program_uuid', 'skill')},
+            },
+        ),
+    ]
```
Comment (DawoudSheraz) [inline taxonomy/migrations/0021_recreate_programskill.py:38]: please add description on PR why this was done.

File: taxonomy/models.py [hunk 2]
```diff
@@ -179,7 +178,6 @@ class ProgramSkill(TimeStampedModel):
         )
     )
     confidence = models.FloatField(
-        blank=False,
```
Comment (DawoudSheraz) [inline taxonomy/models.py:182]: why is this removed?

File: taxonomy/tasks.py [hunk 3]
```diff
@@ -27,3 +27,19 @@ def update_course_skills(course_uuids):
...
+
+@shared_task()
+def update_program_skills(program_uuids):
+    """
+    Task to update program skills.
+
+    Arguments:
+        program_uuids (list): uuids of programs for which skills needs to be updated
+    """
+    LOGGER.info('[TAXONOMY] refresh_program_skills task triggered')
```
Comment (DawoudSheraz) [inline taxonomy/tasks.py:40]: please add the program UUID in the logs.

File: taxonomy/utils.py [hunk 4]
```diff
@@ -144,6 +190,72 @@ def refresh_course_skills(courses, should_commit_to_db):
...
+        'Total Programs Skipped: %s \n'
+        'Total Failures: %s \n',
+        all_failures,
+        success_programs_count,
+        skipped_programs_count,
+        len(all_failures),
+    )
+
+
+def call_emsi_api(text_data, index):
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:240]: nit: get_emsi_skills_data

File: taxonomy/utils.py [hunk 5]
```diff
@@ -56,9 +56,25 @@ def update_skills_data(course_key, skill_external_id, confidence, skill_data):
...
+    skill, __ = Skill.objects.update_or_create(external_id=skill_external_id, defaults=skill_data)
+
+    if not is_program_skill_blacklisted(program_uuid, skill.id):
+        ProgramSkill.objects.update_or_create(
+            program_uuid=program_uuid,
+            skill=skill,
+            defaults={
+                'confidence': confidence,
+            },
+        )
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:72]: It would be good for debugging/info purposes if the result of this is logged.

File: taxonomy/utils.py [hunk 6]
```diff
@@ -144,6 +190,72 @@ def refresh_course_skills(courses, should_commit_to_db):
     )
 
 
+def refresh_program_skills(programs, should_commit_to_db):
+    """
+    Refresh the skills associated with the provided programs.
+    """
+    all_failures = []
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:197]: nit: failures

File: taxonomy/utils.py [hunk 7]
```diff
@@ -144,6 +190,72 @@ def refresh_course_skills(courses, should_commit_to_db):
...
+    LOGGER.info(
+        '[TAXONOMY] Refresh program skills process completed. \n'
+        'Failures: %s \n'
+        'Total Programs Updated Successfully: %s \n'
+        'Total Programs Skipped: %s \n'
+        'Total Failures: %s \n',
+        all_failures,
+        success_programs_count,
+        skipped_programs_count,
+        len(all_failures),
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:236]: please be consistent with string formatting method. You have used f-string method above.

File: tests/test_models.py [hunk 8]
```diff
@@ -111,6 +111,26 @@ def test_string_representation(self):
...
+        """
+        Test the string representation of the ProgramSkill model.
+        """
+        program_skill = factories.ProgramSkillFactory()
+        expected_str = '<ProgramSkill name="{}" program_uuid="{}">'.format(
+            program_skill.skill.name, program_skill.program_uuid
+        )
+        expected_repr = '<ProgramSkill id="{0}" skill="{1!r}">'.format(program_skill.id, program_skill.skill)
+
+        assert expected_str == program_skill.__str__()
```
Comment (DawoudSheraz) [inline tests/test_models.py:130]: you can do str(program_skill) here.

File: tests/test_models.py [hunk 9]
```diff
@@ -111,6 +111,26 @@ def test_string_representation(self):
...
+        assert expected_repr == program_skill.__repr__()
```
Comment (DawoudSheraz) [inline tests/test_models.py:131]: same as str()

File: tests/test_tasks.py [hunk 10]
```diff
@@ -75,3 +76,54 @@ def test_update_course_skills_task_with_no_course_found(self, get_course_skills_
...
+        Verify that `update_program_skills` task work as expected.
+        """
+        get_program_skills_mock.return_value = self.skills_emsi_client_response
+        mock_get_provider.return_value = DiscoveryProgramMetadataProvider([self.program])
+
+        # verify that no `Skill` and `ProgramSkill` records exist before executing the task
+        skill = Skill.objects.all()
+        program_skill = ProgramSkill.objects.all()
+        self.assertEqual(skill.count(), 0)
+        self.assertEqual(program_skill.count(), 0)
```
Comment (DawoudSheraz) [inline tests/test_tasks.py:93]: these lines can be converted into a util method because they are in use across various unit tests.

File: tests/test_utilities.py [hunk 11]
```diff
@@ -140,6 +140,70 @@ def test_update_skills_data(self):
...
+                'type_name': black_listed_program_skill.skill.type_name,
+                'description': black_listed_program_skill.skill.description
+            }
+        )
+
+        updated_name = 'new_name'
+        updated_info_url = 'new_url'
+        updated_type_id = '1'
+        updated_type_name = 'new_type'
+        updated_description = 'new description'
```
Comment (DawoudSheraz) [inline tests/test_utilities.py:166]: nit: these can be converted into dict for better readability.

**Review by DawoudSheraz — COMMENTED** [review]
Left a few comments

File: taxonomy/models.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply taxonomy/models.py:182]: pylint says it can be removed as `blant=False` is the default value.

File: taxonomy/tasks.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply taxonomy/tasks.py:40]: There will be logs for each program_uuid in `refresh_program_skills` so I don't think it is necessary to add them here too.

File: taxonomy/utils.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply taxonomy/utils.py:197]: Already being used.

File: taxonomy/utils.py [hunk 12]
```diff
@@ -104,6 +122,41 @@ def process_skills_data(course, course_skills, should_commit_to_db):
...
+            confidence = float(record['confidence'])
+            skill = record['skill']
+            skill_external_id = skill['id']
+            skill_data = {
+                'name': skill['name'],
+                'info_url': skill['infoUrl'],
+                'type_id': skill['type']['id'],
+                'type_name': skill['type']['name'],
+                'description': skill['description']
+            }
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:146]: nit: this can be converted into a util transform method. The same code is being used in process_course_skills_data.

File: taxonomy/utils.py (hunk 6 above)
Comment (DawoudSheraz) [reply taxonomy/utils.py:212]: You can rename the failure to program_skilll_failures. all_failures is semantically hard to understand.

File: taxonomy/utils.py [hunk 13]
```diff
@@ -157,6 +205,68 @@ def refresh_course_skills(courses, should_commit_to_db):
...
+    LOGGER.info(
+        '[TAXONOMY] Refresh program skills process completed. \n'
+        f'Failures: {all_failures} \n'
+        f'Total Programs Updated Successfully: {success_programs_count} \n'
+        f'Total Programs Skipped: {skipped_programs_count} \n'
+        f'Total Failures: {len(all_failures)} \n',
+    )
+
+
+def get_emsi_skills_data(text_data, index):
```
Comment (DawoudSheraz) [inline taxonomy/utils.py:251]: 👍🏽

**Review by DawoudSheraz — APPROVED** [review]
a few nit comments, otherwise good to do.

**Review by AliAdnanSohail — APPROVED** [review]

**Review by saleem-latif — APPROVED** [review]
Looks Good  👍

File: taxonomy/utils.py [hunk 14]
```diff
@@ -157,6 +205,68 @@ def refresh_course_skills(courses, should_commit_to_db):
...
+    """
+    all_failures = []
+    success_programs_count = 0
+    skipped_programs_count = 0
+
+    for index, program in enumerate(programs, start=1):
+        program_overview = program['overview']
+        if program_overview:
+            try:
+                program_skills = get_emsi_skills_data(program_overview, index)
```
Comment (irfanuddinahmad) [inline taxonomy/utils.py:220]: I think the program skills would be more accurate if they are a union of the skills associated with the courses within the program ... no?

File: taxonomy/utils.py [hunk 15]
```diff
@@ -203,6 +313,24 @@ def is_course_skill_blacklisted(course_key, skill_id):
...
+        program_uuid: uuid of the program whose skill need to be checked.
+        skill_id (int): Primary key identifier of the skill that need to be checked.
+
+    Returns:
+        (bool): True if program-skill (identified by the arguments) is black-listed, False otherwise.
+    """
+    return ProgramSkill.objects.filter(
+        program_uuid=program_uuid,
+        skill_id=skill_id,
+        is_blacklisted=True,
```
Comment (irfanuddinahmad) [inline taxonomy/utils.py:330]: Is there some other flow for the blacklisting i.e. false positives ?

**Review by irfanuddinahmad — APPROVED** [review]
Some comments otherwise LGTM!

File: taxonomy/utils.py (hunk 14 above)
Comment (Ali-D-Akbar) [reply taxonomy/utils.py:220]: We used this approach because of existing edX programs and 2U degrees (which are basically derived from Program). In case of 2U, we will be creating an empty program (i.e. without courses) so taking a union out of it's courses may not be possible in every case.

File: taxonomy/utils.py (hunk 15 above)
Comment (Ali-D-Akbar) [reply taxonomy/utils.py:330]: I haven't written any custom flow for it. The existing util method for course `blacklist_course_skill` isn't being used anywhere so I assumed that the skills can only be blacklisted through django admin.

File: taxonomy/models.py [hunk 16]
```diff
@@ -164,7 +164,6 @@ class ProgramSkill(TimeStampedModel):
     """
 
     program_uuid = models.UUIDField(
-        primary_key=True,
```
Comment (muhammad-ammar) [inline taxonomy/models.py:167]: Why this was removed?

Comment (Ali-D-Akbar) [reply taxonomy/models.py:167]: A program uuid can have more than one skill. It was violating when more than one skill was added against the same uuid. So we removed this constraint and now we're using the autogenerated primary key aka `id`.
